ls1intum/Apollon
# Support for subsystems in component diagrams

### Checklist

### Motivation and Context


As architecture expert in the iPraktikum by the ASE chair, I am using component diagrams to model subsystem decompositions. Unfortunately, Apollon offers no elements other than components making it impossible to create component diagrams following the UML standard. The workaround to reuse components to model subsystems is limited to naming and there has been no way to denote the stereotype ("role") of these elements, e.g. _subsystem_ or _component_, which would be desired in subsystem decompositions and component diagrams in general.

### Description

This change aims to improve the functionality of component diagrams in Apollon. It introduces a new element called "Subsystem", which follows the functionality of a component but has the corresponding stereotype _subsystem_. Furthermore components now also display their stereotype _component_ above their name, following normal component diagrams guidelines. 
Text in subsystems is generally top-aligned, while components use center alignment. Although the styling may be subject to customization for edge cases, such styling should be sufficient for all common use cases. This change deliberately also impacts components in deployment diagrams, as these should follow the same UML standard styling.

### Steps for Testing

1. Run the code and open the website
2. Select "Component Diagram" as the diagram type
3. Drag a subsystem element in the canvas and fill it with a component element
4. Both elements should now display their stereotype correctly and a subsystem should be able to contain a component

### Screenshots


An example of a component diagram that is impossible to create with Apollon without this change:
![Screenshot_463](https://github.com/ls1intum/Apollon/assets/21990230/807416a2-92f1-420b-8163-da6152077c85)
The new sidebar with the changes:
![Screenshot_464](https://github.com/ls1intum/Apollon/assets/21990230/8f3b8388-46fd-42c9-bac6-9b5b87f1c27b)

### Source for UML components / component diagrams
[IBM - Component Diagrams](https://developer.ibm.com/articles/the-component-diagram/)
